examples/chef/common/chef-fan-control-manager.cpp

```diff
@@ -99,98 +124,213 @@ Status ChefFanControlManager::HandleStep(StepDirectionEnum aDirection, bool aWra
-            MatterReportingAttributeChangeCallback(mEndpoint, FanControl::Id, Attributes::SpeedSetting::Id);
-            MatterReportingAttributeChangeCallback(mEndpoint, FanControl::Id, Attributes::SpeedCurrent::Id);
+            DataModel::Nullable<uint8_t> speedSetting(kFanModeLowSpeedRange.Low());
+            SetSpeedSetting(speedSetting);
         }
-
         break;
     }
-    case PercentSetting::Id: {
-        Nullable<uint8_t> newPercentSetting;
-        ReturnErrorOnFailure(aDecoder.Decode(newPercentSetting));
-
-        // Ensure new speed in percent is valid.
-        if (!newPercentSetting.IsNull() && newPercentSetting.Value() > 100)
+    case FanControl::FanModeEnum::kMedium: {
+        if (!kFanModeMediumSpeedRange.Contains(mSpeedCurrent))
         {
-            return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+            DataModel::Nullable<uint8_t> speedSetting(kFanModeMediumSpeedRange.Low());
+            SetSpeedSetting(speedSetting);
         }
-
-        // Only act on changed.
-        if (newPercentSetting != mPercentSetting)
+        break;
+    }
+    case FanControl::FanModeEnum::kOn:
+    case FanControl::FanModeEnum::kHigh: {
+        if (!kFanModeHighSpeedRange.Contains(mSpeedCurrent))
         {
-            mPercentSetting = newPercentSetting;
-
-            // Mark both the setting AND the current dirty, since the current always
-            // tracks the target for our product.
-            MatterReportingAttributeChangeCallback(mEndpoint, FanControl::Id, Attributes::PercentSetting::Id);
-            MatterReportingAttributeChangeCallback(mEndpoint, FanControl::Id, Attributes::PercentCurrent::Id);
+            DataModel::Nullable<uint8_t> speedSetting(kFanModeHighSpeedRange.High());
+            SetSpeedSetting(speedSetting);
         }
-
         break;
     }
-    default:
+    case FanControl::FanModeEnum::kSmart:
+    case FanControl::FanModeEnum::kAuto: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::FanModeWriteCallback: Auto");
         break;
     }
-
-    // Fall through goes to attribute store legacy handling.
-    return CHIP_NO_ERROR;
+    case FanControl::FanModeEnum::kUnknownEnumValue: {
+        ChipLogProgress(NotSpecified, "ChefFanControlManager::FanModeWriteCallback: Unknown");
+        break;
+    }
+    }
 }
 
-CHIP_ERROR ChefFanControlManager::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+void ChefFanControlManager::SetSpeedSetting(DataModel::Nullable<uint8_t> aNewSpeedSetting)
 {
-    VerifyOrDie(aPath.mClusterId == FanControl::Id);
-    VerifyOrDie(aPath.mEndpointId == mEndpoint);
-
-    switch (aPath.mAttributeId)
+    if (aNewSpeedSetting.IsNull())
     {
-    case PercentCurrent::Id: {
-        // Current percents always tracks setting immediately in our implementation.
-        return aEncoder.Encode(mPercentSetting.ValueOr(0));
+        ChipLogError(NotSpecified, "ChefFanControlManager::SetSpeedSetting: invalid value");
```

Comment:
```suggestion
        ChipLogError(NotSpecified, "ChefFanControlManager::SetSpeedSetting: null value is invalid");
```

I think we know explicitly what value is invalid, so the message could be clearer